### PR TITLE
feat(remap): add string comparison capability

### PR DESCRIPTION
**Pull request by jdrouet**
Allow to compare strings

Signed-off-by: Jérémie Drouet <jeremie.drouet@datadoghq.com>

### CODE REVIEW

**Review by JeanMertz — CHANGES_REQUESTED** [review]
Awesome! 🙌  I left one remark.

Other than that, we should update the Cue docs to reflect this change in the language.

File: lib/vrl/compiler/src/expression/op.rs [hunk 1]
```diff
@@ -162,6 +162,15 @@ impl Expression for Op {
             // ... != ...
             Eq | Ne => lhs_def.merge(rhs_def).boolean(),
 
+            // "b" >  "a"
+            // "a" >= "a"
+            // "a" <  "b"
+            // "b" <= "b"
+            Gt | Ge | Lt | Le if lhs_def.is_bytes() && rhs_def.is_bytes() => lhs_def
+                .fallible_unless(K::Bytes)
+                .merge(rhs_def.fallible_unless(K::Bytes))
```
[the remark this answers is not in the document]
Comment (jdrouet) [reply lib/vrl/compiler/src/expression/op.rs:171]: You're right, I did hesitate to put it or not. 👍

File: lib/vrl/compiler/src/expression/op.rs [hunk 2]
```diff
@@ -162,6 +162,14 @@ impl Expression for Op {
             // ... != ...
             Eq | Ne => lhs_def.merge(rhs_def).boolean(),
 
+            // "b" >  "a"
+            // "a" >= "a"
+            // "a" <  "b"
+            // "b" <= "b"
+            Gt | Ge | Lt | Le if lhs_def.is_bytes() && rhs_def.is_bytes() => {
+                lhs_def.scalar(K::Boolean)
```
Comment (JeanMertz) [inline lib/vrl/compiler/src/expression/op.rs:170]: rhs can still be fallible, so we need to merge it with lhs:

```suggestion
                lhs_def.merge(rhs_def).boolean()
```

Comment (JeanMertz) [reply lib/vrl/compiler/src/expression/op.rs:170]: It would be good to add a test that captures this condition.

**Review by JeanMertz — CHANGES_REQUESTED** [review]

**Review by jszwedko — APPROVED** [review]
Nice work! I agree with Jean's remark about fallibility, but otherwise this looks good to me.

File: lib/vrl/compiler/src/expression/op.rs [hunk 3]
```diff
@@ -630,6 +653,11 @@ mod tests {
             want: TypeDef::new().infallible().boolean(),
         }
 
+        less_or_equal_bytes {
+            expr: |_| op(Lt, "bar", "bar"),
```
Comment (JeanMertz) [inline lib/vrl/compiler/src/expression/op.rs:657]:
```suggestion
            expr: |_| op(Le, "bar", "bar"),
```

**Review by JeanMertz — CHANGES_REQUESTED** [review]
One more comment to resolve.

Also, I realized we don't have any integration tests for this change yet. Take a look at [this directory](https://github.com/timberio/vector/tree/master/lib/vrl/tests/tests/expressions/comparison) for more details.

**Review by leebenson — CHANGES_REQUESTED** [review]
This is looking great 👍🏻 

Just a couple of small additions:

1. To add to @JeanMertz's request for tests, I'd specifically like to see comparisons between mixed-case stings and comparing different locales. The immediate use-case outside of VRL is to allow for `{a TO c}` range translations from Datadog search syntax (per https://github.com/timberio/vector/pull/7633), so we should also assert that comparisons work identically to how they would at Datadog.
2. There are a few warnings about transforms not being used. Just an FYI to ensure all tests are passing.

File: lib/vrl/tests/tests/expressions/comparison/ge.vrl [hunk 4]
```diff
@@ -19,4 +19,5 @@
 
     # fallible
     null >= 1 ?? "nope",
+    to_string([.foo]) >= "bar" ?? "nope",
```
Comment (JeanMertz) [inline lib/vrl/tests/tests/expressions/comparison/ge.vrl:22]: We'll also want to fill out the matrix of all other possibilities.

E.g.

```
"a" > "b"
"a" > "a"
"b" > "a"
```

And do that for all comparison operators.

It's a bit of an annoyance to have to write those out manually, especially for the comparison expression, but it's good to cover our bases. (and also, as @leebenson suggested, it's also good to test a few more, such as `1 > "a"` etc).

**Review by JeanMertz — APPROVED** [review]
Sweet! LGTM 🙌

**Comment by jdrouet** [comment]
Regarding the string comparison, datadog is using the unicode string comparison, [same as Rust](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2bb595521481fb63b7394a93fb19f04e).
I think it's good for this question. Regarding the CI warnings, after rebasing, it's fixed.

**Review by leebenson — APPROVED** [review]
🎉

**Pull request merged**